src/Microsoft.AspNetCore.Http/PipeWriterAdapter.cs

```diff
@@ -0,0 +1,287 @@
+                // and flush the result.
+                if (_completedSegments != null)
+                {
+                    var count = _completedSegments.Count;
+                    for (var i = 0; i < count; i++)
+                    {
+                        var segment = _completedSegments[i];
+                        await _writingStream.WriteAsync(segment.Buffer, 0, segment.Length, joinedToken);
+                    }
+                }
+
+                if (_currentSegment != null)
+                {
+                    await _writingStream.WriteAsync(_currentSegment, 0, _position, joinedToken);
+                }
+
+                await _writingStream.FlushAsync(joinedToken);
+            }
+            catch (OperationCanceledException)
+            {
+                // Catch any cancellation and translate it into setting isCanceled = true
+                return new FlushResult(isCanceled: true, IsCompletedOrThrow());
+            }
+
+            // After writing to the stream, we can return all ArrayPool segments that are completed.
+            ReturnCompletedBuffers();
+
+            return new FlushResult(isCanceled: false, IsCompletedOrThrow());
+        }
+
+        private void EnsureCapacity(int sizeHint)
+        {
+            // This does the Right Thing. It only subtracts _position from the current segment length if it's non-null.
+            // If _currentSegment is null, it returns 0.
+            var remainingSize = _currentSegment?.Length - _position ?? 0;
+
+            // If the sizeHint is 0, any capacity will do
+            // Otherwise, the buffer must have enough space for the entire size hint, or we need to add a segment.
+            if ((sizeHint == 0 && remainingSize > 0) || (sizeHint > 0 && remainingSize >= sizeHint))
+            {
+                // We have capacity in the current segment
+                return;
+            }
+
+            AddSegment(sizeHint);
+        }
+
+        private void AddSegment(int sizeHint = 0)
+        {
+            if (_currentSegment != null)
+            {
+                // We're adding a segment to the list
+                if (_completedSegments == null)
+                {
+                    _completedSegments = new List<CompletedBuffer>();
+                }
+
+                // Position might be less than the segment length if there wasn't enough space to satisfy the sizeHint when
+                // GetMemory was called. In that case we'll take the current segment and call it "completed", but need to
+                // ignore any empty space in it.
+                _completedSegments.Add(new CompletedBuffer(_currentSegment, _position));
+            }
+
+            // Get a new buffer using the minimum segment size, unless the size hint is larger than a single segment.
+            _currentSegment = ArrayPool<byte>.Shared.Rent(Math.Max(_minimumSegmentSize, sizeHint));
```

Comment:
Hmm not sure we want that in the default path.

Reply:
`ArrayPool<byte>.Shared` devirtualizes now https://github.com/dotnet/coreclr/pull/20637 so if you were using `IMemoryPool` you'd want something like
```csharp
var size = Math.Max(_minimumSegmentSize, sizeHint);
_currentSegment = _pool?.AllocByte(size) ?? ArrayPool<byte>.Shared.Rent(size);
```

Though, isn't `IMemoryPool` in the wrong namespace? (i.e. downstream)

Reply:
IMemoryPool allocates like a mofo, we can avoid it in this case with little downside.

Reply:
Why don't we do the same in pipe then?

Reply:
Or think about the way to fix that allocation

Reply:
There are some concrete benefits you get from the IMemoryPool
1. Somebody can plug in an implementation (it's an abstraction)
2. We have an implementation (in ASP.NET Core) that uses pre-pinned memory (a variant of the first one)

The default implementation (ArrayMemoryPool) allocates a managed object per segment. It's not that big a deal when doing large allocations but can be avoided completely when using the array pool. We get no benefit in ASP.NET Core if these Streams use pre-pinned buffers since the Streams we'll be writing to won't be pinning buffer for async reads. That happens in the transport layer in Kestrel.

Before reverting to use the IMemoryPool, we should get the baseline so we can measure the differences.